GitHub: use GitHub actions to close stale issues and PRs

jbech-linaro opened the pull request:
It's not uncommon that people ask questions or send pull requests that
becomes stale for various reasons. Some examples can be:
 - Original Poster (OP) doesn't close the ticket.
 - OP doesn't get back when maintainers asked for more information.
 - OP gets busy with other things and have to put the work a side.

Having lots of open issues and pull request makes it almost impossible
to figure out where help and input is needed, since things simply
becomes overwhelming. By using GitHub actions [1] we enable a tool that
can:
  - Automatically close issues that no one cares about.
  - Help us bring attention back to issues that are of importance.
  - Push patch creators to not drop their proposed patches.
  - Remind maintainers and other OP-TEE users that reviews are needed.

This patch set propose actions that will warn after 30 days of no
activity, together with that warning the GitHub bot will add a "stale"
label. If that label isn't removed within 5 days, then the GitHub bot
will close the issue or pull request.

People who are admins in the OP-TEE project can remove the label by
themselves, other people have to add a comment telling that their issue
is still relevant and that they would like to have the stale label
removed.

When we first enable this, hundreds of issues and pull requests will be
affected, but after a first round the list of open issues and pull
request should be much smaller and should only contain things that are
or relevance to continue working with.

[1] https://help.github.com/en/actions/automating-your-workflow-with-github-actions/about-github-actions

Signed-off-by: Joakim Bech <joakim.bech@linaro.org>

jforissier commented on .github/workflows/stale_issue.yml:
```diff
@@ -0,0 +1,15 @@
+    - uses: actions/stale@v1
+      with:
+        repo-token: ${{ secrets.GITHUB_TOKEN }}
+        stale-issue-message: 'This has issue has been marked as a stale issue because it has been open (more than) 30 days with no activity. Remove the stale label or add a comment saying that you would like to have the label removed otherwise this issue will automatically be closed in 5 days. Note, that you can always re-open a closed issue at any time.'
```
`s/This has issue/This issue/`

jforissier commented on .github/workflows/stale_pr.yml:
```diff
@@ -0,0 +1,15 @@
+    - uses: actions/stale@v1
+      with:
+        repo-token: ${{ secrets.GITHUB_TOKEN }}
+        stale-pr-message: 'This has pull request has been marked as stale because it has been open (more than) 30 days with no activity. Remove the stale label or add a comment saying that you would like to have the label removed otherwise this pull request will automatically be closed in 5 days. Note, that you can always re-open a closed pull request at any time.'
```
`s/This has pull request/This pull request/`

jbech-linaro commented:
> I think this is a good idea. 

:+1: 

> Does the 30 days delay restart again when we remove the stale flag?

Yes it does. To give an example, have a look at [this](https://github.com/jbech-linaro/ibart/issues/9) IBART issue where I played around with it. 28th November I manually removed the stale label, 29th December the bot/action re-added it automatically and January 4th the bot/action closed it automatically.

> Two minor things below, other than that:

Will fix, thanks!

jbech-linaro commented:
Squashed, rebased and tag(s) applied (should be) ready for merge! Thanks! And once merged, I'll take care for the rest of configuration which only can be done by admins.

jforissier commented:
The checkpatch error and warning could be fixed easily...

jbech-linaro commented:
Oh, my bad for not noticing that. Another fix will be sent in a couple of minutes.

jbech-linaro commented:
After fixing the obvious error, I'm down to these and I have no idea if and how to resolve them. Do you have any idea or suggestions?

```bash
WARNING: Possible unwrapped commit description (prefer a maximum 75 chars per line)
#36:
[1] https://help.github.com/en/actions/automating-your-workflow-with-github-actions/about-github-actions

WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#48:
new file mode 100644

total: 0 errors, 2 warnings, 30 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

0001-GitHub-use-actions-to-close-stale-issues-and-PRs.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.
```

jenswi-linaro commented:
```
Link: [1] https://help.github.com/en/actions/automating-your-workflow-with-github-actions/about-github-actions
```

jenswi-linaro commented:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

jbech-linaro commented:
> Link: [...

Thanks Jens, it works. Interesting solution, it complains about a long line, solution .. make it longer :)

However, the last one. 
```
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#48:
new file mode 100644
```
I've tried Google it, but I cannot find an answer how to silence that. So either we simply ignore it (my vote goes here) or I have to add a section in the MAINTAINERS for it.

jenswi-linaro commented:
I think we should ignore it. It's the same problem in the Linux kernel by the way. I think this warning is more of a reminder, annoying though.

jbech-linaro commented:
> annoying though.

Indeed ... new patch has been pushed.

jbech-linaro commented:
Hepps! A bit surprised to see that Travis/checkpatch doesn't complain about the "maintainer warning", but I'm not sorry for that :) Maybe I have a more recent version of checkpatch locally?

Second try ... Squashed, rebased and tag(s) applied (should be) ready for merge! Thanks!

jforissier commented:
> Hepps! A bit surprised to see that Travis/checkpatch doesn't complain about the "maintainer warning", but I'm not sorry for that :) Maybe I have a more recent version of checkpatch locally?

This warning should be suppressed by `--ignore FILE_PATH_CHANGES`, which is set [here](https://github.com/OP-TEE/optee_os/blob/master/scripts/checkpatch_inc.sh#L21). How did you run checkpatch locally?

jbech-linaro commented:
> How did you run checkpatch locally?

Since it doesn't change any "code" I just run it directly on the patch (`git format-patch -1 && ./linux/scripts/checkpatch.pl <mypatch>`. Thanks for merging! Should be interesting to see when this triggers, free up you inbox! :)

jforissier commented:
> > How did you run checkpatch locally?
> 
> Since it doesn't change any "code" I just run it directly on the patch (`git format-patch -1 && ./linux/scripts/checkpatch.pl <mypatch>`.

Ah OK, that explains why. You could have just run `./scripts/checkpatch.sh HEAD` ;-)

> Thanks for merging! Should be interesting to see when this triggers, free up you inbox! :)

Yeah, email storm expected, not a big deal!
